content/blog/20200130-jepsen.md

```diff
@@ -0,0 +1,81 @@
+
+### 1. A simple core
+
+A simple solution to a problem space usually results in a robust system. As an example, etcd's Raft implementation is one of the simplest implementations in open source. It focuses on the core state machine of Raft and avoids wall-time and I/O handling in the core library. And, in its API, etcd supports multi-key transactions, but adopts a simple transactional model to keep the system easier to comprehend.
+
+### 2. Rigorous testing practice
+
+Early in the development cycle of etcd, we set a goal to [achieve 80% unit test line coverage](https://github.com/etcd-io/etcd/issues/1467) and 90% unit test line coverage over core components like Raft and the MVCC storage engine. As the project progressed, we put effort into ensuring logical path coverage for core components by [introducing fail points](https://github.com/etcd-io/gofail). To embrace the randomness of timing, OS scheduling and asynchronous network I/Os, we also developed continuous integration tests, which run 24x7, to try and produce issues on a running system in a way that unit tests simply cannot. These tests have increased the quality of etcd significantly.
+
+## Identified Issues
+
+Of course, no system undergoes Jepsen testing without them reporting some areas where it could be better.  etcd had a few:
+
+### Lock Issues
+
+One lock implementation issue was found during the test where etcd failed to check the ownership of the lock before the pending lock API call returned.
+
+More specifically, in etcd a lock acquirer is associated with a session. Only if the session is not expired is the acquirer considered to still hold the lock. When acquiring the lock, the acquirer might not get the lock immediately if the lock is held by someone else. In that case the etcd server will put the acquirer into a queue and waits for its turn before returning the lock API call. However, the session might expire during this wait time. So even if it is the acquirer's turn to get the lock, the etcd server must check the existence of the session again before returning to the API call.
+
+### Documentation Issues
+
+The Jepsen analysis does not only cover software correctness but also verifies the documentation claims. The Jepsen author found documentation issues and misleading use case examples of etcd.  The previous etcd documentation on consistency and isolation is not clear and might cause confusion.
+
+It describes the consistency model based on [the Wikipedia definition](https://en.wikipedia.org/wiki/Consistency_model), which is also the classification some etcd engineers [learned in University](http://www.cs.cmu.edu/~srini/15-446/S09/lectures/10-consistency.pdf). The previous documentation also separates the isolation level from consistency level, as there is no general agreement on [how the two should be defined together](https://fauna.com/blog/demystifying-database-systems-part-4-isolation-levels-vs-consistency-levels).
```

Comment:
It's not fully clear what "It" is here given the paragraph break. I'd replace with "That documentation describes the consistency model..."